File: abc070d.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <utility>
#include <stack>
#include <queue>

using namespace std;

typedef pair<int, int> Pii;
typedef long long int ll;

int main() {
  cin.tie(0);
  ios::sync_with_stdio(false);

  int n;
  cin >> n;
  vector<vector<pair<int, ll> > > edge(n);
  for (int i = 0; i < n-1; i++) {
    int a, b, c;
    cin >> a >> b >> c;
    edge[a-1].emplace_back(b-1, c);
    edge[b-1].emplace_back(a-1, c);
  }
  int q, k;
  cin >> q >> k;
  k--;
  vector<int> x(q), y(q);
  for (int i = 0; i < q; i++) {
    cin >> x[i] >> y[i];
    x[i]--;
    y[i]--;
  }

  vector<ll> distanceToK(n);
  vector<bool> visited(n);
  queue<pair<int, ll> > que;
  que.emplace(k, 0);
  while (!que.empty()) {
    auto now = que.front();
    que.pop();
    if (visited[now.first]) continue;
    visited[now.first] = true;
    distanceToK[now.first] = now.second;
    for (auto &z: edge[now.first]) {
      if (!visited[z.first]) que.emplace(z.first, now.second + z.second);
    }
  }

  vector<ll> ans(q);
  for (int i = 0; i < q; i++) ans[i] = distanceToK[x[i]] + distanceToK[y[i]];

  for (auto &z: ans) cout << z << endl;

  return 0;
}

File: abc070d.py

def main():
    n = int(input())
    edge = []
    for i in range(n):
        edge.append([])
    for i in range(n-1):
        abc = list(map(int, input().split()))
        edge[abc[0]-1].append((abc[1]-1, abc[2])) # zero index
        edge[abc[1]-1].append((abc[0]-1, abc[2])) # zero index
    q, k = map(int, input().split())
    k -= 1 # zero index
    xy = []
    for i in range(q):
        xy.append(list(map(int, input().split())))
        xy[-1][0] -= 1 # zero index
        xy[-1][1] -= 1 # zero index

    distance_from_k = []
    visited = []
    for i in range(n):
        distance_from_k.append(0)
        visited.append(False)
    st = [k]
    while st:
        now = st.pop()
        if visited[now]:
            continue
        visited[now] = True
        for x in edge[now]:
            if visited[x[0]]:
                continue
            distance_from_k[x[0]] = distance_from_k[now] + x[1]
            st.append(x[0])

    ans = []
    for i in range(q):
        ans.append(distance_from_k[xy[i][0]] + distance_from_k[xy[i][1]])

    for x in ans:
        print(x)

if __name__ == '__main__':
    main()
